### PR TITLE
Add check_preconditions flag to PauliSum and PauliString expectation functions

### LINKED ISSUE #2118: Add PauliSum-level expectation from density matrix without validation
If a user is confident their density matrix is valid, it would be good if they could calculate the expectation value without the additional overhead of an eigenvalue decomposition or numerical error causing this computation to error out (especially when the density matrix represents a nominally pure state). @peterse thoughts?

### PR DESCRIPTION
**Pull request by kevinsung**
Fixes #2118

### CODE REVIEW

File: cirq/ops/pauli_string.py [hunk 1]
```diff
@@ -395,7 +398,9 @@ def _expectation_from_wavefunction_no_validation(
     def expectation_from_density_matrix(self,
                                         state: np.ndarray,
                                         qubit_map: Mapping[raw_types.Qid, int],
-                                        atol: float = 1e-7) -> float:
+                                        atol: float = 1e-7,
```
Comment (Strilanc) [inline cirq/ops/pauli_string.py]: While you're at it, force keywords starting with `atol`.

Add `check_preconditions` to the `Args:` list.

File: cirq/ops/pauli_string.py [hunk 2]
```diff
@@ -303,7 +303,9 @@ def _apply_unitary_(self, args: 'protocols.ApplyUnitaryArgs'):
     def expectation_from_wavefunction(self,
                                       state: np.ndarray,
                                       qubit_map: Mapping[raw_types.Qid, int],
-                                      atol: float = 1e-7) -> float:
+                                      atol: float = 1e-7,
```
Comment (Strilanc) [inline cirq/ops/pauli_string.py]: While you're at it, force keywords starting with `atol`.

Add `check_preconditions` to the `Args:` list.

File: cirq/ops/linear_combinations.py [hunk 3]
```diff
@@ -376,17 +378,20 @@ def expectation_from_wavefunction(self, state: np.ndarray,
...
         return sum(
             p._expectation_from_wavefunction_no_validation(state, qubit_map)
             for p in self)
 
-    def expectation_from_density_matrix(self, state: np.ndarray,
-                                        qubit_map: Mapping[raw_types.Qid, int]
+    def expectation_from_density_matrix(self,
+                                        state: np.ndarray,
+                                        qubit_map: Mapping[raw_types.Qid, int],
+                                        check_preconditions: bool = True
```
Comment (Strilanc) [inline cirq/ops/linear_combinations.py]: Force keyword arg on `check_preconditions`. Add `check_preconditions` to the `Args:` list.

File: cirq/ops/linear_combinations.py [hunk 4]
```diff
@@ -342,8 +342,10 @@ def copy(self) -> 'PauliSum':
         factory = type(self)
         return factory(self._linear_dict.copy())
 
-    def expectation_from_wavefunction(self, state: np.ndarray,
-                                      qubit_map: Mapping[raw_types.Qid, int]
+    def expectation_from_wavefunction(self,
+                                      state: np.ndarray,
+                                      qubit_map: Mapping[raw_types.Qid, int],
+                                      check_preconditions: bool = True
```
Comment (Strilanc) [inline cirq/ops/linear_combinations.py]: Force keyword arg on `check_preconditions`. Add `check_preconditions` to the `Args:` list.

**Comment by kevinsung** [comment]
Done

**Pull request merged**